### PR TITLE
objects: add a history tracker and have forms use the `useHistoryTracker` hook

### PR DESCRIPTION
**Pull request by tahini**
This is a first step towards removing the `ObjectWithHistory` from the class hierarchy. Tracking history is not required for most use cases except when editing from the UI. Also, the current way of tracking history was to save history after each field set, resulting in non-atomic updates (see #9), requiring multiple undo/redo to really go back to previous state.

We add a `HistoryTracker` class which saves an object's attributes history. This tracker is used in the `useHistoryTracker` hook that forms can use to manage value changes (a step toward #48). The form decide when it should save the history, after all attribute updates have been done.

The `TransitRoutingForm` has been updated to use this new approach, as it does not concern a saveable object, making a good first candidate to implement the new approach. This actually adds the undo/redo functionality to the TransitRouting form, that did not previously exist.

## Summary by CodeRabbit

* **New Features**
  * Added immutable undo/redo history tracker, a React hook to track form/object changes, and undo/redo UI buttons for user-facing controls.
  * Integrated history-based undo/redo into the transit routing form for reverting/restoring form and object state.

* **Refactor**
  * Centralized form state management around the new history mechanism for simpler, more reliable updates.

* **Tests**
  * Added comprehensive unit tests for history behavior, hook integration, undo/redo flows, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/chaire-lib-frontend/src/components/forms/useHistoryTracker.tsx [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            // Reset the invalidFields ref when object ID changes
+            invalidFieldsRef.current = {};
+            setInvalidFieldsCount(0);
+
+            // Also reset the ref when object ID changes
+            formValuesRef.current = _cloneDeep(object.attributes);
+            setFormValues(_cloneDeep(object.attributes));
+        },
+        (object as any).id !== undefined ? [(object as any).id] : []
+    );
```
[the remark this answers is not in the document]
Comment (tahini) [reply packages/chaire-lib-frontend/src/components/forms/useHistoryTracker.tsx]: No this is fine actually. Since undo/redo creates a new object, the suggestion will cause this effect to re-run whenever those functions are executed, effectively allowing only one undo/redo per object without ID. In the code base, object classes that have an ID field will always have this field defined. But in case some object has an actually undefined id that can be defined later, it's safe to use `[(object as any).id]` for all cases: object without ids will never retrigger the effect and if the ID is eventually set, it will retrigger.
